src/Microsoft.AspNet.OData.Shared/EdmChangedObjectCollectionOfT.cs

```diff
@@ -0,0 +1,297 @@
+                            operation = DataModificationOperationKind.Unlink;
+                        }
+
+                        if (originalObj != null)
+                        {
+                            //This case handle deletions
+                            original.Remove(originalObj);
+                        }
+                    }
+                    else
+                    {
+                        if (originalObj == null)
+                        {
+                            operation = DataModificationOperationKind.Insert;
+                            //This case handle additions
+                            originalObj = Activator.CreateInstance(changedObj.ExpectedClrType);
+                            original.Add(originalObj);
+                        }
+
+                        //Patch for addition/update. This will call Delta<T> for each item in the collection
+                        changedObj.Patch(originalObj);
+                    }
+
+                    edmChangedObject = changedObj;
+                }
+                catch
+                {
+                    //For handling the failed operations.
+                    HandleFailedOperation(changedObj, operation, ref edmChangedObject, originalObj);
+                }
+
+                edmChangedObjectCollection.Add(edmChangedObject);
+            }
+
+            return edmChangedObjectCollection;
+        }
+               
+
+        private void HandleFailedOperation(dynamic changedObj, DataModificationOperationKind operation, ref IEdmChangedObject edmChangedObject, TStructuralType originalObj)
+        {
+            DataModificationExceptionType dataModificationExceptionType = new DataModificationExceptionType(operation);
+
+            // This handles the Data Modification exception. This adds Core.DataModificationException annotation and also copy other instance annotations.
+            //The failed operation will be based on the protocol
+            switch (operation)
+            {
+                case DataModificationOperationKind.Update:
+                    edmChangedObject = changedObj;
+                    break;
+                case DataModificationOperationKind.Insert:
+                    {
+                        Type type = typeof(EdmDeltaDeletedEntityObject<>).MakeGenericType(changedObj.ExpectedClrType);
+                        EdmDeltaDeletedEntityObject edmDeletedObject  = Activator.CreateInstance(type, EntityType) as EdmDeltaDeletedEntityObject;
+                        object contentId = changedObj.TransientInstanceAnnotationContainer.GetResourceAnnotation("Core.ContentID");
+                        if(contentId != null)
+                        {
+                            edmDeletedObject.Id = contentId.ToString();                            
+                        }
+                        else
+                        {
+                            edmDeletedObject.Id = string.Empty;
+                        }
```

Comment:
I'm not sure on this logic.  The id of the deleted (failed insert) of the resource is not necessarily (likely not) the value of the Core.ContentID annotation. The id should be copied from the incoming changedObj (along with any other properties, especially key properties) to the edmDeletedObject. ODL will generate the id from the key values, if supplied, so only if neither an id nor key values are supplied should we try to set the Id of the deleted object.  And then, yes, if we copy values over to the ODataDeletedResource and call Get on the .Id and still get null, we can set it to the ContentID or empty string.